### ISSUE TITLE
Add a scraping OJAD data source

### ISSUE THREAD

**Issue by Mindful** [Mindful/cardbuilder#8]
[OJAD](http://www.gavo.t.u-tokyo.ac.jp/ojad/pages/notes) is an excellent resource for Japanese pronunciation data, including actual audio. Their licensing basically just says
> 教育・学術研究を目的とした活動に限って利用することができます。営利活動を目的とした企業の研究開発での利用はお控えください。

So it doesn't seem like using their data would be an issue. They don't have an API though, so the data will need to be scraped.

**Comment by huntingb** [comment]
I have been watching this repo for a while and I see that you are beginning to work on the OJAD data source. I started exploring the site in a Jupyter Notebook that I am going to attach. I found a way to download the audio files at least of the dictionary forms of words and get pitch accent data from the table cells in a form like this:
[('は', 0), ('た', 1), ('ら', 1), ('く', 1)]
0 for low, 1 for high for each mora. It would also be cool to be able to download the drawing of the accent curve that's above each form, but this is in a canvas element and I'm not sure how to scrape from those. I haven't yet tried to reverse engineer the JS that's drawing them, if it is even on the client side. I am still trying to figure out how all the pieces in your project fit together so I can start contributing to it but maybe some of this code will help you in the meantime.

[Link to notebook
](https://github.com/huntingb/ojad-notebook/blob/master/scraping.ipynb)

**Comment by Mindful** [comment]
@huntingb 
This is excellent, thank you! I'm traveling so my availability is pretty limited this week, but I will look at this in more depth when I get back. I started in on an OJAD data source [here](https://github.com/Mindful/cardbuilder/blob/main/cardbuilder/lookup/ja/ojad.py) although there's no meaningful content yet, so I'd be happy to have it filled in.

In regard to contributing in general: there's nowhere near enough documentation right now. The readme is out of date too, which doesn't help anything. Documentation is on my to-do list though and I am looking for serious contributors, so if you are interested, I'd be happy to walk you through your first contributions sometime next week or the week after. Understanding what's confusing for you will also help me figure out what needs to be documented. 

In regards to OJAD: OJAD is run by the University of Tokyo which happens to be where I'm currently doing research, so I've been meaning to just reach out to the lab that owns it and see if they'd be willing to for example add an API. Worst case they might just pass along some info about how to draw the accent curve -- I 100% agree that would be great to have. I will try and get this done sooner rather than later.

**Comment by Mindful** [comment]
Alright, a few updates.

1. I looked through the notebook in more depth - content in there could definitely be used to make a scraping data source. I can of course do it, but I think it would be a great first contribution if you wanted to do it. The way you're using BeautifulSoup in the notebook is fine by the way, don't worry about copying what I did in ScrapingMerriamWebster unless you find it easier that way. 
2. In regards to reverse engineering JS for the canvas elements - I wouldn't worry about this yet. It's possible you might just be able to cheat and use something like Selenium to draw the canvas and download that, but even just having pitch accent and audio data for words would be a huge win, so I think it's safe to start there.
3. I reached out to the lab that runs OJAD and they've stated they would like to work on an API, but work won't start until this October at the earliest. So for now scraping is definitely the way to do this.

Finally, while it's still a work in progress, I have started writing documentation and drafted a [data source implementation guide](https://cardbuilder.readthedocs.io/en/latest/contributing/contrib_data_source.html). This is unlikely to answer all your questions, but it should hopefully provide somewhere to start.
